career-mentorship-program/2021-spring/matching_code.py

```diff
@@ -410,9 +503,66 @@ def compute_weight_and_statistics(self):
         if year_difference == 1 or year_difference == 2:
             weight += MENTOR_CLOSE_IN_YEAR_BONUS
 
-
         """ !!! Add any new matching criteria here !!! """
 
+        # 4. Hobbies
+
+        num_hobbies_overlap = len(self.mentee.hobbies & self.mentor.hobbies)
+        num_mentee_hobbies = len(self.mentee.hobbies)
+        hobbies_overlap_fraction = num_hobbies_overlap / num_mentee_hobbies
+        statistics["hobbies_overlap_fraction"] = hobbies_overlap_fraction
+        weight += HOBBIES_TOPICS_MULTIPLIER * hobbies_overlap_fraction
+
+        # function for gender and race
+        def gender_and_race(mentor_preference, mentee_preference, stat_field1, stat_field2, bonus):
+            if mentor_preference and mentee_preference:
+                does_prefer = True
+                statistics[stat_field1] = True
+            else:
+                does_prefer = False
+                statistics[stat_field1] = False
+            if does_prefer:
+                preferred_and_same = self.mentor.actual == self.mentee.actual
+                statistics[stat_field2] = preferred_and_same
+            else:
+                preferred_and_same = False
+                statistics[stat_field2] = preferred_and_same
+            if does_prefer and preferred_and_same:
+                self.weight += bonus
+
+        # 5. Gender
+        # checks if both want to be matched with same gender and are same gender, if so add bonus
+        gender_and_race(self.mentor.does_prefer_gender, self.mentee.does_prefer_gender, 'does_prefer_gender',
+                        'gender_preferred_and_same', GENDER_SAME_BONUS)
+
+        # 6. Race
+        # checks if both want to be matched with same race and are same race, if so add bonus
+        gender_and_race(self.mentor.does_prefer_race, self.mentee.does_prefer_race, 'does_prefer_race',
```

Comment:
Note that race is represented as a set, so the calculation is actually different than for gender. A suggestion of how to compute the weight is in [this comment](https://github.com/alaynarichmond/northwestern-cs-mentorship/pull/3/files#r604616057).